### PR TITLE
Support dir stores with string/Buffer content type to handle non textual static files SALTO-850, SALTO-815

### PR DESCRIPTION
**Pull request by omrilit**

### CODE REVIEW

**Comment by omrilit** [comment]
This PR is rebased on top of https://github.com/salto-io/salto/pull/1258

File: packages/core/src/app_config.ts [hunk 1]
```diff
@@ -133,7 +133,7 @@ const mergeConfigWithEnv = async (config: AppConfig): Promise<AppConfig> => {
 }
 
 const configFromNaclFile = async (filepath: string): Promise<AppConfig> => {
-  const buf = await readFile(filepath)
+  const buf = await readFile(filepath) as Buffer
```
Comment (ori-moisis) [inline packages/core/src/app_config.ts]: this cast should not be needed

File: packages/core/src/local-workspace/dir_store.ts [hunk 2]
```diff
@@ -31,19 +28,20 @@ const WRITE_CONCURRENCY = 100
 const DELETE_CONCURRENCY = 100
 const RENAME_CONCURRENCY = 100
 
-type FileMap = {
-  [key: string]: dirStore.File
+type FileMap<T extends dirStore.ContentType> = {
+  [key: string]: dirStore.File<T>
 }
```
Comment (ori-moisis) [inline packages/core/src/local-workspace/dir_store.ts:32]: nit:
```suggestion
type FileMap<T extends dirStore.ContentType> = Record<string, dirStore.File<T>>
```

File: packages/core/src/local-workspace/dir_store.ts [hunk 3]
```diff
@@ -72,41 +70,42 @@ const buildLocalDirectoryStore = (
...
-    return await exists(absFileName)
+    return await fileUtils.exists(absFileName)
       ? {
         filename,
-        buffer: await readTextFile(absFileName),
-        timestamp: (await stat(absFileName) as Stats).mtimeMs,
-      }
+        buffer: await fileUtils.readFile(absFileName, encoding),
+        timestamp: (await fileUtils.stat(absFileName) as Stats).mtimeMs,
+      } as dirStore.File<T>
```
Comment (ori-moisis) [inline packages/core/src/local-workspace/dir_store.ts]: would be better to cast only `buffer` instead of the whole object IMO

File: packages/hubspot-adapter/test/change_validators/readonly.test.ts [hunk 4]
```diff
@@ -64,13 +64,6 @@ describe('readonly change validator', () => {
     })
 
     describe('keep resolved value of name in ContactProperty', () => {
-      it('should not have an error when replacing name with static file with same value', async () => {
-        after.value.name = new StaticFile({ filepath: 'path', content: Buffer.from(contactPropertyMock.name) })
-        changeErrors = await readonlyValidator([
-          toChange({ before: contactPropertyInstance, after }),
-        ])
-      })
```
Comment (ori-moisis) [inline packages/hubspot-adapter/test/change_validators/readonly.test.ts:72]: seems unrelated to this PR, why was this changed? and why only the test changed?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 5]
```diff
@@ -206,14 +206,17 @@ const buildNaclFilesSource = (
     return (await getState()).elementsIndex[topLevelID.parent.getFullName()] || []
   }
 
+  const getNaclFile = async (filename: string): Promise<NaclFile | undefined> =>
+    naclFilesStore.get(filename)
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:210]: nit: I don't think this is a positive change, it adds more text and complexity (the reader of the code now has to jump to a different function to see implementation that was clear enough before this change) for no apparent reason

File: packages/workspace/src/workspace/static_files/source.ts [hunk 6]
```diff
@@ -34,15 +34,15 @@ class LazyStaticFile extends StaticFile {
     if (this.internalContent === undefined) {
       const file = this.dirStore.getSync(this.filepath)
       if (file !== undefined) {
-        this.internalContent = Buffer.from(file.buffer)
+        this.internalContent = file.buffer
       }
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/static_files/source.ts:38]: nit:
```
this.internalContent = this.dirStore.getSync(this.filepath)?.buffer
```

File: packages/hubspot-adapter/test/change_validators/readonly.test.ts (hunk 4 above)
Comment (omrilit) [reply packages/hubspot-adapter/test/change_validators/readonly.test.ts:72]: Written in the commit message
> remove a UT that expects the resolved value of a static file is a string (and not buffer)

Comment (ori-moisis) [reply packages/hubspot-adapter/test/change_validators/readonly.test.ts:72]: either the test needed to be fixed or the validator needs to change, but it doesn't make sense that because static files now hold buffers we no longer need to test something that was important enough to write a test about.
I am missing something in my logic? how is this related?

Comment (omrilit) [reply packages/hubspot-adapter/test/change_validators/readonly.test.ts:72]: This test is not really a UT and it calls resolveValues within it.
It used the staticFile (assuming it resolves to string) in order to check that it is the same value. It's just wrong from now on and it's checked for others. I can modify the expected result but I think that it's redundant and the test can be omitted. Don't you think?

Comment (ori-moisis) [reply packages/hubspot-adapter/test/change_validators/readonly.test.ts:72]: the test says:
`should not have an error when replacing name with static file with same value`
you are saying that after your change to the code, it will have an error when "replacing name with static file with the same value", right?
that means you caused a behavior change in that validator, the test caught this behavior change, so, the test did its job, because, whether we like it or not, @adamro wanted the validator to not have an error in this case.
if you think we should really have an error in this case, well, we'll have to dive deeper into why Adam explicitly wanted otherwise

File: packages/workspace/src/workspace/static_files/source.ts [hunk 7]
```diff
@@ -34,7 +34,7 @@ class LazyStaticFile extends StaticFile {
     if (this.internalContent === undefined) {
       const file = this.dirStore.getSync(this.filepath)
       if (file !== undefined) {
-        this.internalContent = file.buffer
+        this.internalContent = this.dirStore.getSync(this.filepath)?.buffer
```
Comment (ori-moisis) [inline packages/workspace/src/workspace/static_files/source.ts:37]: we should not read the file twice, my nit was that the whole `if (file !== undefined)` part is redundant, it wasn't super important before but this version really should not be merged

Comment (omrilit) [reply packages/workspace/src/workspace/static_files/source.ts:37]: Sure!
just forgot removing the above one... thanks!

**Comment by ori-moisis** [comment]
please add jira ticket reference
